rllib/algorithms/algorithm.py

```diff
@@ -941,6 +943,199 @@ def duration_fn(num_units_done):
+            return batch, metrics, w_seq_no
+
+        rollout_metrics = []
+
+        # How many episodes have we run (across all eval workers)?
+        num_units_done = 0
+        round_ = 0
+        time_started = time.time()
+        timed_out = True
+
+        while time.time() - time_started < self.config["evaluation_sample_timeout_s"]:
+            units_left_to_do = duration_fn(num_units_done)
+            if units_left_to_do <= 0:
+                timed_out = False
+                break
+
+            round_ += 1
+            # Use the AsyncRequestsManager to get ready evaluation results and
+            # metrics.
+            self._evaluation_async_req_manager.call_on_all_available(
+                remote_fn=remote_fn,
+                fn_args=[weights_ref, self._evaluation_weights_seq_number],
+            )
+            ready_requests = self._evaluation_async_req_manager.get_ready()
+
+            batches = []
+            for i, requests in enumerate(ready_requests.values()):
+                for req in requests:
+                    batch, metrics, seq_no = req
+                    # Ignore results, if the weights seq-number does not match (is from
+                    # a previous evaluation step) OR if we have already reached the
+                    # configured duration (e.g. number of episodes to evaluate for).
+                    if (
+                        seq_no == self._evaluation_weights_seq_number
+                        and (
+                            i * (1 if unit == "episodes" else rollout_len * num_envs)
+                            < units_left_to_do
+                        )
+                    ):
+                        batches.append(batch)
+                        rollout_metrics.extend(metrics)
+
+            _agent_steps = sum(b.agent_steps() for b in batches)
+            _env_steps = sum(b.env_steps() for b in batches)
+
+            # 1 episode per returned batch.
+            if unit == "episodes":
+                num_units_done += len(batches)
+                # Make sure all batches are exactly one episode.
+                for ma_batch in batches:
+                    ma_batch = ma_batch.as_multi_agent()
+                    for batch in ma_batch.policy_batches.values():
+                        assert np.sum(batch[SampleBatch.DONES])
+            # n timesteps per returned batch.
+            else:
+                num_units_done += (
+                    _agent_steps if self._by_agent_steps else _env_steps
+                )
+            if self.reward_estimators:
+                all_batches.extend(batches)
+
+            agent_steps_this_iter += _agent_steps
+            env_steps_this_iter += _env_steps
+
+            logger.info(
+                f"Ran round {round_} of parallel evaluation "
+                f"({num_units_done}/{duration if not auto else '?'} "
+                f"{unit} done)"
+            )
+
+        if timed_out and log_once("evaluation_timeout"):
```

Comment:
we can get rid of this warning for async mode.

Reply:
done, no more timeouts in async mode